### PR TITLE
perf(sql): improve performance of queries with negative limits and existing order by clauses

### PR DESCRIPTION
**Pull request by nwoolmer**
Fixes https://github.com/questdb/questdb/issues/5138
Fixes https://github.com/questdb/questdb/issues/4198

**Changes**

- rename `PageFrameFwdRowCursorFactory -> FwdPageFrameRowCursorFactory` for consistency with `BwdPageFrameRowCursorFactory`
- Add condition in SqlCodeGenerator to check worth the order by clause starts with a desc designated timestamp and is limited.
- in some cases, push down the limit from an outer select-choose model to inner select-none (i.e when there is an appropriate order by on the select-none and we want to generate `LimitedSizePartiallySortedRecordCursorFactory` (for 4198)
- in some cases, push down an appropriate reverse scan to handle a negative limit efficiently, and create an outer sort to preserve the intended query order (for 5138)
- add new `allowPropagationofOrderByAdvice` flag to prevent order by advice existing on parent models from overriding advice in lower models. This should only really be employed when trying to force a particular fwd/bwd scan on a select-none node.

**Benchmark**

Schema:

```sql
CREATE TABLE 'bench' (
  symbol SYMBOL capacity 256 CACHE,
  timestamp TIMESTAMP,
  price DOUBLE,
  amount DOUBLE
) timestamp (timestamp) PARTITION BY DAY WAL;
````

100 million rows, M1 Macbook Pro 16 GB RAM.

```
-- 5138
 SELECT timestamp, price FROM bench ORDER BY timestamp ASC, price DESC LIMIT -3;
-- master: 2.47 seconds
-- pr: 300 micros

-- 4198
 SELECT timestamp, price FROM bench ORDER BY timestamp DESC, price DESC LIMIT 3;
-- master: 3.06 seconds
-- pr: 295 micros
```

**For [5138](https://github.com/questdb/questdb/issues/5138):**

Take this query, with an order by clause starting with the designated timestamp, and a negative limit.

```sql
EXPLAIN SELECT timestamp, side FROM trades ORDER BY timestamp ASC, side DESC LIMIT -3;
```
with plan:
```
Sort light lo: -3 partiallySorted: true
  keys: [timestamp, side desc]
    PageFrame
        Row forward scan
        Frame forward scan on: trades
```

This query fails on demo with an error:

```
limit of 33554432 memory exceeded in LimitedSizeLongTreeChain
```

With this change, the query is rewritten, and compiles to this plan:

```
Sort light
  keys: [timestamp, side desc]
    Sort light lo: 3 partiallySorted: true
      keys: [timestamp desc, side desc]
        PageFrame
            Row backward scan
            Frame backward scan on: trades
```

**For [4198](https://github.com/questdb/questdb/issues/4198)**

Here is a query with multiple order by clauses, starting with designated timestamp in descending order:

```sql
SELECT cab_type, vendor_id, pickup_datetime from trips 
ORDER BY pickup_datetime DESC, cab_type DESC 
LIMIT 100;
```

This produces the following plan:

```sql
Sort light lo: 100
  keys: [pickup_datetime desc, cab_type desc]
    PageFrame
        Row forward scan
        Frame forward scan on: trips
```

In this case, the database scans the whole table, sorts it, and then takes the last 100 rows. This fails the same way as the first example:

```
limit of 33554432 memory exceeded in LimitedSizeLongTreeChain
```

Now, it will produce the following plan:

```sql
Sort light lo: 100 partiallySorted: true
  keys: [pickup_datetime desc, cab_type desc]
    PageFrame
        Row backward scan
        Frame backward scan on: trips
```

### CODE REVIEW

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 40 / 40 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|33|33|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 48 / 48 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|38|38|100.00%|

**Comment by ideoma** [comment]
Heyj,

I'm not sure this is correct plan for the query `SELECT timestamp, side FROM trades ORDER BY timestamp ASC, side DESC LIMIT -3`

```
Sort light
  keys: [timestamp, side desc]
    Limit lo: 3
        PageFrame
            Row backward scan
            Frame backward scan on: trades
```

the limit should apply after the sort, not before.

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 95 / 96 (98.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|72|73|98.63%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FwdPageFrameRowCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|17|17|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 1]
```diff
@@ -4194,6 +4256,75 @@ private boolean rewriteNegativeLimitGuard(QueryModel model, SqlExecutionContext
...
+                    nested.getOrderByDirectionAdvice().add(nested.getOrderByDirection().get(i));
+                }
+            } else {
+                // assume already filled
+                nested.getOrderByDirectionAdvice().set(0, ORDER_DIRECTION_DESCENDING);
+            }
+
+            nested.setAllowPropagationOfOrderByAdvice(false); // stop propagation
+
+            // copy the integral part i.e if its -3, the 3
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:4320]: nit: a typo here?
```suggestion
            // copy the integral part i.e if it's -3, then 3
```

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 2]
```diff
@@ -4194,6 +4256,75 @@ private boolean rewriteNegativeLimitGuard(QueryModel model, SqlExecutionContext
...
+            // we want to push down a limited reverse scan, and then sort into the intended ordering afterwards
+
+            // first, copy the order by up
+            for (int i = 0, n = nested.getOrderBy().size(); i < n; i++) {
+                model.addOrderBy(nested.getOrderBy().get(i), nested.getOrderByDirection().get(i));
+                model.getOrderByAdvice().add(nested.getOrderBy().get(i));
+                model.getOrderByDirectionAdvice().add(nested.getOrderByDirection().get(i));
+
+                // also strip advice
+                model.getOrderByAdvice().remove(i);
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:4301]: What are we doing here? Are we removing older order by advice elements? If so, is it guaranteed that `model.getOrderByAdvice()`'s size matches the `nested.getOrderBy()`'s one?

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java [hunk 3]
```diff
@@ -646,7 +659,8 @@ public boolean equals(Object o) {
                 && Objects.equals(unionModel, that.unionModel)
                 && Objects.equals(updateTableModel, that.updateTableModel)
                 && Objects.equals(updateTableToken, that.updateTableToken)
-                && skipped == that.skipped;
+                && skipped == that.skipped
+                && allowPropagationOfOrderByAdvice == that.allowPropagationOfOrderByAdvice;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/model/QueryModel.java:663]: Could you modify `QueryModelTest` to verify this check, as well as the `clear()` method modification?

File: core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java [hunk 4]
```diff
@@ -1901,6 +1907,20 @@ public void testOrderingOfSortsInSingleTimestampCase() throws Exception {
         });
     }
 
+    @Test
+    public void testPushDownLimitFromChooseToNone() throws Exception {
+        assertMemoryLeak(() -> {
+            ddl(tripsDdl);
+            drainWalQueue();
+            assertPlanNoLeakCheck("select cab_type, vendor_id, pickup_datetime from trips \n" +
+                    "order by pickup_datetime desc, cab_type desc limit 100;", "Sort light lo: 100 partiallySorted: true\n" +
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java:1916]: Would be great to add tests with non-null hi limit here and below just to make sure we didn't break anything there.

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 96 / 97 (98.97%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|73|74|98.65%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FwdPageFrameRowCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|17|17|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 5]
```diff
@@ -3577,6 +3579,60 @@ private void pushDownOrderByAdviceToJoinModels(QueryModel model, ObjList<QueryMo
...
+                        && model.getUnionModel() == null
+                        && model.getJoinModels().size() == 1
+                        && model.getGroupBy().size() == 0
+                        && model.getSampleBy() == null
+                        && !hasAggregateQueryColumn(model)
+                        && !model.isDistinct()
+                        && nested.getTimestamp() != null
+                        && Chars.equalsIgnoreCase(nested.getTimestamp().token, nested.getOrderBy().get(0).token)
+                        && nested.getOrderByDirection().get(0) == ORDER_DIRECTION_DESCENDING
+                        && (loFunction = getLoFunction(model.getLimitLo(), executionContext)) != null
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/SqlOptimiser.java:3620]: Thanks, there are lots of checks, so I missed it. Could you co-locate the checks, so that they read better?

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 94 / 95 (98.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|71|72|98.61%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FwdPageFrameRowCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|17|17|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 94 / 95 (98.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|71|72|98.61%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FwdPageFrameRowCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|17|17|100.00%|

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java [hunk 6]
```diff
@@ -133,6 +145,7 @@ public class QueryModel implements Mutable, ExecutionModel, AliasTranslator, Sin
     // be the only fields for these values.
     public ExpressionNode fillTimestampAlias;
     private ExpressionNode alias;
+    private boolean allowPropagationOfOrderByAdvice = true;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/model/QueryModel.java:148]: nit: would be great to add a comment with the purpose of this field

**Review by puzpuzpuz — APPROVED** [review]

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 7]
```diff
@@ -4194,6 +4259,78 @@ private boolean rewriteNegativeLimitGuard(QueryModel model, SqlExecutionContext
...
+     * This would ordinarily compile to a `LimitedSizePartiallySortedRecordCursor`.
+     * That means it will forward scan - and run out of memory on demo!
+     * Instead, we aim to subquery a backwards scan and sort.
+     * This then needs to be compiled into a `LimitedSizePartiallySortedRecordCursor`, but in reverse
+     * in order to correctly handle duplicates.
+     * <p>
+     * So we produce a model like this:
+     * `select-choose timestamp, side from (trades timestamp (timestamp) order by timestamp desc, side desc limit 3) order by timestamp, side desc`
+     */
+    private void rewriteNegativeLimitWithOrderBy(QueryModel model) throws SqlException {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:4277]: exception not thrown

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 8]
```diff
@@ -5258,7 +5266,7 @@ public boolean supportsUpdateRowId(TableToken tableToken) {
             final boolean intervalHitsOnlyOnePartition;
             if (intrinsicModel.hasIntervalFilters()) {
                 RuntimeIntrinsicIntervalModel intervalModel = intrinsicModel.buildIntervalModel();
-                if (orderDescendingByDesignatedTimestampOnly) {
+                if (orderDescendingByDesignatedTimestampOnly || orderByStartsWithDescDesignatedTimestampAndLimited) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:5269]: these two variables are always used together, it could one variable

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 9]
```diff
@@ -4194,6 +4256,78 @@ private boolean rewriteNegativeLimitGuard(QueryModel model, SqlExecutionContext
...
+
+            int firstOrderByDir = nested.getOrderByDirection().get(0);
+
+            if (firstOrderByDir != ORDER_DIRECTION_ASCENDING) {
+                return;
+            }
+
+            ExpressionNode firstOrderByArg = nested.getOrderBy().get(0);
+
+            if (!Chars.equals(firstOrderByArg.token, nested.getTimestamp().token)) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:4295]: this should be case-insensitive + test

File: core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java [hunk 10]
```diff
@@ -1901,6 +1907,34 @@ public void testOrderingOfSortsInSingleTimestampCase() throws Exception {
...
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java:1916]: nit: code formatting makes the test hard to read. When args are arranged vertically, please do not have two or more args on the same line

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 11]
```diff
@@ -4194,6 +4256,78 @@ private boolean rewriteNegativeLimitGuard(QueryModel model, SqlExecutionContext
...
+                return;
+            }
+
+            // we want to push down a limited reverse scan, and then sort into the intended ordering afterwards
+
+            // first, copy the order by up
+            for (int i = 0, n = nested.getOrderBy().size(); i < n; i++) {
+                model.addOrderBy(nested.getOrderBy().get(i), nested.getOrderByDirection().get(i));
+                model.getOrderByAdvice().add(nested.getOrderBy().get(i));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:4304]: copying order advice is unnecessary when order is also copied

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 112 / 117 (95.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/ObjList.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|85|88|96.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FwdPageFrameRowCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|17|17|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 113 / 118 (95.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/ObjList.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|85|88|96.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FwdPageFrameRowCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|18|18|100.00%|

**Review by bluestreak01 — APPROVED** [review]
